pkg/datasource/formatters/json/json.go

```diff
@@ -0,0 +1,333 @@
+				e.Write(fieldName)
+				e.Write([]byte("{"))
+			})
+			fns = append(fns, subFieldFuncs...)
+			fns = append(fns, func(e *encodeState, data datasource.Data) {
+				e.Write([]byte("}"))
+			})
+			continue
+		}
+
+		var fn func(e *encodeState, data datasource.Data)
+		// Field doesn't have subfields
+		switch accessor.Type() {
+		case api.Kind_Int16:
+			fn = func(e *encodeState, data datasource.Data) {
+				b := strconv.AppendInt(e.scratch[:0], int64(int16(f.ds.ByteOrder().Uint16(accessor.Get(data)))), 10)
+				e.Write(b)
+			}
+		case api.Kind_Int32:
+			fn = func(e *encodeState, data datasource.Data) {
+				b := strconv.AppendInt(e.scratch[:0], int64(int32(f.ds.ByteOrder().Uint32(accessor.Get(data)))), 10)
+				e.Write(b)
+			}
+		case api.Kind_Int64:
+			fn = func(e *encodeState, data datasource.Data) {
+				b := strconv.AppendInt(e.scratch[:0], int64(f.ds.ByteOrder().Uint16(accessor.Get(data))), 10)
+				e.Write(b)
+			}
+		case api.Kind_Uint16:
+			fn = func(e *encodeState, data datasource.Data) {
+				b := strconv.AppendUint(e.scratch[:0], uint64(f.ds.ByteOrder().Uint16(accessor.Get(data))), 10)
+				e.Write(b)
+			}
+		case api.Kind_Uint32:
+			fn = func(e *encodeState, data datasource.Data) {
+				b := strconv.AppendUint(e.scratch[:0], uint64(f.ds.ByteOrder().Uint32(accessor.Get(data))), 10)
+				e.Write(b)
+			}
+		case api.Kind_Uint64:
+			fn = func(e *encodeState, data datasource.Data) {
+				b := strconv.AppendUint(e.scratch[:0], f.ds.ByteOrder().Uint64(accessor.Get(data)), 10)
+				e.Write(b)
+			}
+		case api.Kind_Float32:
+			fn = func(e *encodeState, data datasource.Data) {
+				floatEncoder(32).writeFloat(e, float64(math.Float32frombits(f.ds.ByteOrder().Uint32(accessor.Get(data)))))
+			}
+		case api.Kind_Float64:
+			fn = func(e *encodeState, data datasource.Data) {
+				floatEncoder(32).writeFloat(e, math.Float64frombits(f.ds.ByteOrder().Uint64(accessor.Get(data))))
+			}
+		case api.Kind_String:
+			fn = func(e *encodeState, data datasource.Data) {
+				writeString(e, string(accessor.Get(data)))
+			}
+		case api.Kind_Bool:
+			fn = func(e *encodeState, data datasource.Data) {
+				// handle arbitrary length bools
+				for b := range accessor.Get(data) {
+					if b != 0 {
+						e.WriteString("true")
+						return
+					}
+				}
+				e.WriteString("false")
+			}
+		default:
+			fn = func(e *encodeState, data datasource.Data) {
+				writeString(e, gadgets.FromCString(accessor.Get(data)))
+			}
+		}
+		fns = append(fns, func(e *encodeState, data datasource.Data) {
+			e.Write(fieldName)
+			fn(e, data)
+		})
+	}
+	return
+}
+
+func (f *Formatter) Marshal(data datasource.Data) []byte {
+	e := bufpool.Get().(*encodeState)
+	e.Reset()
+	defer bufpool.Put(e)
+	for _, fn := range f.fns {
+		fn(e, data)
+	}
+	return e.Bytes()
+}
+
+type floatEncoder int // number of bits
+
+// from encoding/json/encode.go
+func (bits floatEncoder) writeFloat(e *encodeState, f float64) {
```

Comment:
Go's package uses reflection which we can avoid (or: can't use ourselves in the first place here). However, their internal functions that serialize strings/floats are proven, stable & valid in terms of the JSON spec, so I'd rather keep them instead of re-implementing it - and in a way that is performant. Sadly it's not exported.

Reply:
I haven't found one, otherwise I would have.